##### Importing Modules

In [10]:
from data_transform import DataTransform
from dataframe_info import DataFrameInfo

##### Loading in the data

In [11]:
df_info = DataFrameInfo("loans.csv")
df_info.show_rows(num_rows=4, head_or_tail='tail')

Unnamed: 0      id  member_id  loan_amount  funded_amount  \
54227       54227  142608      74724         5000         5000.0   
54228       54228  117045      70978         3500         3500.0   
54229       54229   88854      70699         5000         5000.0   
54230       54230   72323      70694         5000         5000.0   

       funded_amount_inv       term  int_rate  instalment grade sub_grade  \
54227             2350.0  36 months     10.59      162.73     C        C2   
54228             2225.0  36 months      7.43      108.77     A        A2   
54229              225.0  36 months      7.43      155.38     A        A2   
54230              350.0  36 months     11.54      164.98     C        C5   

      employment_length home_ownership  annual_inc verification_status  \
54227          < 1 year           RENT     15000.0        Not Verified   
54228         10+ years       MORTGAGE    300000.0        Not Verified   
54229           4 years           RENT    200000.0        Not Verified   
54230           9 years       MORTGAGE     55000.0        Not Verified   

      issue_date                                        loan_status  \
54227   Oct-2013                                         Fully Paid   
54228   Aug-2013  Does not meet the credit policy. Status:Fully ...   
54229   Aug-2013                                         Fully Paid   
54230   Jul-2013  Does not meet the credit policy. Status:Charge...   

      payment_plan             purpose    dti  delinq_2yrs  \
54227            n               other  24.40            0   
54228            n  debt_consolidation   1.00            0   
54229            n               house   0.28            0   
54230            n               other  10.00            0   

      earliest_credit_line  inq_last_6mths  mths_since_last_delinq  \
54227             Jan-1999               3                     0.0   
54228             Feb-1984               0                     NaN   
54229             Mar-1995               0                     0.0   
54230             Oct-1998               4                     NaN   

       mths_since_last_record  open_accounts  total_accounts  out_prncp  \
54227                     0.0              6               6        0.0   
54228                     NaN              8              18        0.0   
54229                     0.0              2               2        0.0   
54230                     NaN              6              13        0.0   

       out_prncp_inv  total_payment  total_payment_inv  total_rec_prncp  \
54227            0.0        5931.25            2787.47          5000.00   
54228            0.0        3915.37            2489.01          3500.00   
54229            0.0        5174.18             232.84          5000.00   
54230            0.0        1501.19             104.60           966.58   

       total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
54227         826.31          104.942381        0.00                     0.00   
54228         415.37            0.000000        0.00                     0.00   
54229         174.18            0.000000        0.00                     0.00   
54230         352.34            0.000000      182.27                     1.84   

      last_payment_date  last_payment_amount next_payment_date  \
54227          Oct-2016               490.01               NaN   
54228          Sep-2016               110.58          Sep-2016   
54229          Mar-2014                 0.00               NaN   
54230          Mar-2014               164.97          Dec-2014   

      last_credit_pull_date  collections_12_mths_ex_med  \
54227              Sep-2016                         0.0   
54228              May-2013                         NaN   
54229              May-2013                         NaN   
54230              Jan-2015                         NaN   

       mths_since_last_major_derog  policy_code application_type  
54227                          NaN     

There are some columns where I would like to view unique categorical values in particular columns such as `loan_status` or `purpose`.

In [12]:
df_info.count_categorical('loan_status')

loan_status
Fully Paid                                             27037
Current                                                19268
Charged Off                                             5571
Does not meet the credit policy. Status:Fully Paid       984
Late (31-120 days)                                       580
Does not meet the credit policy. Status:Charged Off      368
In Grace Period                                          265
Late (16-30 days)                                        106
Default                                                   52
Name: count, dtype: int64

In [13]:
df_info.count_categorical('purpose')

purpose
debt_consolidation    29593
credit_card           11133
other                  3697
home_improvement       3215
major_purchase         1679
small_business         1312
car                    1022
medical                 631
wedding                 539
moving                  456
house                   354
vacation                320
educational             201
renewable_energy         79
Name: count, dtype: int64

Above shows the most common reasons for taking out a loan, debt consolidation being the most popular.

#### Formatting columns

Before formatting the columns, we need to view the data in each column first. We can do this by using the info method.

In [14]:
df_info.get_info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54231 entries, 0 to 54230
Data columns (total 44 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Unnamed: 0                   54231 non-null  int64  
 1   id                           54231 non-null  int64  
 2   member_id                    54231 non-null  int64  
 3   loan_amount                  54231 non-null  int64  
 4   funded_amount                51224 non-null  float64
 5   funded_amount_inv            54231 non-null  float64
 6   term                         49459 non-null  object 
 7   int_rate                     49062 non-null  float64
 8   instalment                   54231 non-null  float64
 9   grade                        54231 non-null  object 
 10  sub_grade                    54231 non-null  object 
 11  employment_length            52113 non-null  object 
 12  home_ownership               54231 non-null  object 
 13  annual_inc      

Here we can see that the data across all columns mostly consists of int, float, object and some null values.
This dataset contains 54231 rows and 44 columns.

The columns that appear to contain null values are the following: `funded_amount`, `term`, `int_rate`, `employment_length`, `mths_since_last_delinq`, `mths_since_last_record`, `last_payment_date`, `next_payment_date`, `last_credit_pull_date`, `collections_12_mths_ex_med` and `mths_since_last_major_derog`.

We will deal with the null values later but for now lets change the data types of columns that aren't suitable.

In [15]:
df_info.describe_columns()

Unnamed: 0            id     member_id   loan_amount  funded_amount  \
count  54231.000000  5.423100e+04  5.423100e+04  54231.000000   51224.000000   
mean   27115.000000  7.621797e+06  8.655350e+06  13333.076100   13229.509117   
std    15655.285561  9.571362e+06  1.031281e+07   8082.196709    8019.017599   
min        0.000000  5.552100e+04  7.069400e+04    500.000000     500.000000   
25%    13557.500000  7.594330e+05  9.587720e+05   7000.000000    7000.000000   
50%    27115.000000  7.084590e+06  8.709873e+06  12000.000000   12000.000000   
75%    40672.500000  8.860616e+06  1.052714e+07  18000.000000   18000.000000   
max    54230.000000  3.867612e+07  4.146185e+07  35000.000000   35000.000000   

       funded_amount_inv      int_rate    instalment    annual_inc  \
count       54231.000000  49062.000000  54231.000000  5.423100e+04   
mean        12952.622979     13.507328    400.013953  7.222085e+04   
std          8099.473527      4.392893    238.920012  5.158934e+04   
min             0.000000      5.420000     15.670000  3.300000e+03   
25%          6700.000000     10.370000    224.205000  4.500000e+04   
50%         11300.000000     13.160000    347.150000  6.100000e+04   
75%         18000.000000     16.200000    527.550000  8.600000e+04   
max         35000.000000     26.060000   1407.010000  2.039784e+06   

                dti   delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
count  54231.000000  54231.000000    54231.000000             23229.00000   
mean      15.867087      0.241799        0.893843                34.01920   
std        7.623124      0.706706        1.261421                21.92324   
min        0.000000      0.000000        0.000000                 0.00000   
25%       10.200000      0.000000        0.000000                15.00000   
50%       15.600000      0.000000        1.000000                31.00000   
75%       21.260000      0.000000        1.000000                49.00000   
max       39.910000     18.000000       33.000000               146.00000   

       mths_since_last_record  open_accounts  total_accounts     out_prncp  \
count             6181.000000   54231.000000    54231.000000  54231.000000   
mean                74.829154      10.559809       24.168206   2753.964172   
std                 36.137264       4.797169       11.587025   5154.893646   
min                  0.000000       1.000000        1.000000      0.000000   
25%                 53.000000       7.000000       16.000000      0.000000   
50%                 87.000000      10.000000       23.000000      0.000000   
75%                103.000000      13.000000       31.000000   3397.325000   
max                119.000000      53.000000      108.000000  32399.940000   

       out_prncp_inv  total_payment  total_payment_inv  total_rec_prncp  \
count   54231.000000   54231.000000       54231.000000     54231.000000   
mean     2752.568147   12079.208480       11788.946618      9407.048589   
std      5153.043643    8391.043845        8363.508506      6958.124264   
min         0.000000       0.000000           0.000000         0.000000   
25%         0.000000    5895.210000        5658.815000      4264.350000   
50%         0.000000   10113.250000        9835.830000      7644.920000   
75%      3394.980000   16272.895000       15978.200000     12505.625000   
max     32399.940000   55758.222084       55061.000000     35000.020000   

       total_rec_int  total_rec_late_fee    recoveries  \
count   54231.000000        54231.000000  54231.000000   
mean     2577.757101            0.901512     93.501288   
std      2581.657345            6.215792    630.843636   
min         0.000000            0.000000      0.000000   
25%       889.560000            0.000000      0.000000   
50%      1734.640000            0.000000      0.000000   
75%      3323.765000            0.000000      0.000000   
max     23062.450000          358.680000  27750.000000   

       collection_recovery_fee  last_payment_amount  \
count 

This shows a statistical overview of all columns in the dataset.

In [16]:
df_info.null_percentages()

Empty DataFrame
Columns: [columns, null_percent]
Index: []

# Data Transformations

#### Converting column data types

First, I will change the column containing categorical data. Though it may make sense to convert these columns to strings, changing it to a "category" data type will utilise less memory.

In [17]:
category_columns = [
    'term',
    'grade',
    'sub_grade',
    'employment_length',
    'home_ownership',
    'verification_status',
    'loan_status',
    'purpose'
]

In [18]:
transform_df.to_categorical(category_columns)
transform_df.show_rows(1)

NameError: name 'transform_df' is not defined

Now we need to convert all of the date columns to the correct format.

In [ ]:
date_columns = [
    'issue_date',
    'earliest_credit_line',
    'last_payment_date',
    'next_payment_date',
    'last_credit_pull_date'
]

for date_column in date_columns:
    transform_df.format_dates(date_column)

transform_df.get_info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54231 entries, 0 to 54230
Data columns (total 44 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Unnamed: 0                   54231 non-null  int64         
 1   id                           54231 non-null  int64         
 2   member_id                    54231 non-null  int64         
 3   loan_amount                  54231 non-null  int64         
 4   funded_amount                51224 non-null  float64       
 5   funded_amount_inv            54231 non-null  float64       
 6   term                         49459 non-null  category      
 7   int_rate                     49062 non-null  float64       
 8   instalment                   54231 non-null  float64       
 9   grade                        54231 non-null  category      
 10  sub_grade                    54231 non-null  category      
 11  employment_length            52113 non-nu

We can now see that the data type for the formatted date columns is now `datetime64[ns]` and is formatted into yyyy-mm-dd.

All that is left now is to change `term` into a numerical value.

In [ ]:
transform_df.term_to_numeric("term")
transform_df.show_rows(3)

Unnamed: 0        id  member_id  loan_amount  funded_amount  \
0           0  38676116   41461848         8000         8000.0   
1           1  38656203   41440010        13200        13200.0   
2           2  38656154   41439961        16000        16000.0   

   funded_amount_inv  term  int_rate  instalment grade sub_grade  \
0             8000.0  36.0      7.49      248.82     A        A4   
1            13200.0  36.0      6.99      407.52     A        A3   
2            16000.0  36.0      7.49      497.63     A        A4   

  employment_length home_ownership  annual_inc verification_status issue_date  \
0           5 years       MORTGAGE     46000.0        Not Verified 2021-01-01   
1           9 years           RENT     50000.0        Not Verified 2021-01-01   
2           8 years       MORTGAGE     73913.0     Source Verified 2021-01-01   

  loan_status payment_plan      purpose    dti  delinq_2yrs  \
0     Current            n  credit_card  19.54            2   
1     Current            n  credit_card  24.20            0   
2  Fully Paid            n  credit_card  16.92            0   

  earliest_credit_line  inq_last_6mths  mths_since_last_delinq  \
0           1987-10-01               1                     5.0   
1           2001-09-01               0                     NaN   
2           1998-09-01               0                    69.0   

   mths_since_last_record  open_accounts  total_accounts  out_prncp  \
0                     NaN             12              27    5529.70   
1                     NaN             15              31    9102.83   
2                     NaN              7              18       0.00   

   out_prncp_inv  total_payment  total_payment_inv  total_rec_prncp  \
0        5529.70        2982.51            2982.51          2470.30   
1        9102.83        4885.11            4885.11          4097.17   
2           0.00       16824.54           16824.54         16000.00   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         512.21                 0.0         0.0                      0.0   
1         787.94                 0.0         0.0                      0.0   
2         824.54                 0.0         0.0                      0.0   

  last_payment_date  last_payment_amount next_payment_date  \
0        2022-01-01               248.82        2022-02-01   
1        2022-01-01               407.52        2022-02-01   
2        2021-10-01             12850.16               NaT   

  last_credit_pull_date  collections_12_mths_ex_med  \
0            2022-01-01                         0.0   
1            2022-01-01                         0.0   
2            2021-10-01                         0.0   

   mths_since_last_major_derog  policy_code application_type  
0                          5.0            1       INDIVIDUAL  
1                          NaN            1       INDIVIDUAL  
2                          NaN            1       INDIVIDUAL

#### Getting information from the dataframe

In [ ]:
from dataframe_info import DataFrameInfo

Text

In [ ]:
transform_df